File: SimulationCpp/projektPO/roslina.cpp
#include "roslina.h"
#include <iostream>
using namespace std;

void Roslina::akcja() {
	if (probaRozsiania()) {
		Organizm* n = dziecko();
		rozsiej(n);
	}
}

void Roslina::rozsiej( Organizm* n) {
	int nX = x, nY = y, g = BRAKMIEJSC, r;
	int t2[4][2];
	for (int i = SASIADGORA; i < SASIADDOL; i++) {
		if (x + i < POCZATEKPLANSZY || x + i >= s->getXM()) continue;
		for (int j = SASIADLEWO; j < SASIADPRAWO; j++) {
			if (y + j < POCZATEKPLANSZY || y + j >= s->getYM()) continue;

			if ((i == SRODEK && j != SRODEK) || (i != SRODEK && j == SRODEK))

				if (!s->tab[x + i][y + j]) {
					t2[g][XWOLNE] = x + i;
					t2[g][YWOLNE] = y + j;
					g++;
				}
		}
	}
	if (g != BRAKMIEJSC) {
		r = rand() % g;
		nX = t2[r][XWOLNE];
		nY = t2[r][YWOLNE];
	}

	if (s->tab[nX][nY] == nullptr) {
		n->setX(nX);
		n->setY(nY);
		Organizm* obecny = this;
		while (obecny->getNas() != nullptr) {
			obecny = obecny->getNas();
		}
		n->zycie = MLODY;
		n->setPop(obecny);
		obecny->setNas(n);
		s->tab[nX][nY] = n;
		
	}
	else delete[] n;
}

bool Roslina::probaRozsiania() {
	int i = rand() % SZANSANAROZSIA;
	if (i == UDANA)return true;
	return false;
}

void Roslina::kolizja(int nX, int nY) {
	zycie = MARTWY;
	makeLog(s->tab[nX][nY]->wypiszNazwe(), wypiszNazwe(), s->log, to_string(x), to_string(y), KOMr2);
	s->tab[x][y] = s->tab[nX][nY];
	s->tab[nX][nY] = nullptr;
	s->tab[x][y]->setX(x);
	s->tab[x][y]->setY(y);
}

void Roslina::makeLog(string z1, string z2, string& log, string x, string y, int akcja) {
	if (akcja == KOMr2)log += z1 + " zjada " + z2 +" na pozycji (" + x + ", " + y + ")" + "\n";
	if (akcja == KOMr3)log += z1 + " zjada " + z2 + " i zwieksza swoja sile o 3 na pozycji (" + x + ", " + y + ")" + "\n";
	if (akcja == KOMr4)log += z1 + " zjada " + z2 + " i umiera na pozycji (" + x + ", " + y + ")" + "\n"; 
	if (akcja == KOMr5)log += z1 + " zabija " + z2 + " na pozycji (" + x + ", " + y + ")" + "\n";
}

Roslina::~Roslina() {
}

File: SimulationJava/src/Roslina.java
import java.util.Random;

public abstract class Roslina  extends Organizm {


    protected boolean probaRozsiania() {
        Random rand = new Random();
        int i = rand.nextInt(10);
        if (i == 0)return true;
        return false;
    }

    protected void rozsiej(Organizm n){
        int nX = x, nY = y,g= 0,r;
        int[] t2 = new int[s.getIleS()];
        int[][] t=s.getSasiad(this);
        for(int i=0;i<s.getIleS();i++){
            if(t[i][0]!=-1){
                t2[g]=i;
                g++;
            }
        }
        
        if (g != 0) {
            Random rand = new Random();
            r = rand.nextInt(g);
            nX = t[t2[r]][0];
            nY = t[t2[r]][1];
        }
        if(s.getPole(nX, nY)==null){
            n.setX(nX);
            n.setY(nY);
            n.setZycie(2);
            s.dodaj(nX, nY, n);
            makeLog(wypiszNazwe(), "", 1, nX, nY);
        }
    }

    public void makeLog(String A, String B, int co,int x,int y){
        if(co==1) s.setLog(s.getLog() + A +" rozsiewa sie na: (" + x + "," + y + ")" + "\n");
        if(co==2) s.setLog(s.getLog() + A +" zjada "+ B +" na: (" + x + "," + y + ")" + "\n");
        if(co==3) s.setLog(s.getLog() + A +" zjada "+ B +" i umiera na: (" + x + "," + y + ")" + "\n");
        if(co==4) s.setLog(s.getLog() + A +" zjada "+ B +" i zwieksza sile na: (" + x + "," + y + ")" + "\n");
        if(co==5) s.setLog(s.getLog() + A +" zabija "+ B +" na: (" + x + "," + y + ")" + "\n");
    }

    @Override
    protected void kolizja(int xx,int yy) {
        zycie = 0;
        s.setTab(x, y, s.getPole(xx, yy));
        s.setTab(xx,yy,null);
        s.getPole(x, y).setX(x);
        s.getPole(x, y).setY(y);
        makeLog(s.getPole(x, y).wypiszNazwe(), wypiszNazwe(), 2, x, y);
    }

    @Override
    public void akcja() {
       if(probaRozsiania()){
           Organizm n = dziecko();
           rozsiej(n);
       }
    }


     
}
